crates/router/src/core/payouts.rs

```diff
@@ -228,45 +227,69 @@ pub async fn payouts_cancel_core(
     if helpers::is_payout_terminal_state(status) {
         return Err(report!(errors::ApiErrorResponse::InvalidRequestData {
             message: format!(
-                "{} is in terminal state [status: {}]",
-                payout_create.payout_id.to_owned(),
-                status
+                "Payout {} cannot be cancelled for status {}",
+                payout_data.payout_create.payout_id, status
             ),
         }));
-    }
 
-    // TODO: Remove hardcoded connector
-    let connector_name = api_enums::Connector::Adyen;
+    // Make local cancellation
+    } else if helpers::is_eligible_for_local_payout_cancellation(status) {
+        let updated_payout_create = storage::PayoutCreateUpdate::StatusUpdate {
+            connector_payout_id: connector_payout_id.to_owned(),
+            status: storage_enums::PayoutStatus::Cancelled,
+            error_message: Some("Cancelled by user".to_string()),
+            error_code: None,
+            is_eligible: None,
+        };
+        payout_data.payout_create = state
+            .store
+            .update_payout_create_by_merchant_id_payout_id(
+                &merchant_account.merchant_id,
+                &payout_create.payout_id,
+                updated_payout_create,
+            )
+            .await
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("Error updating payout_create in db")?;
 
-    // Form connector data
-    let connector_data: api::ConnectorData = api::ConnectorData::get_connector_by_name(
-        &state.conf.connectors,
-        &connector_name.to_string(),
-        api::GetToken::Connector,
-    )
-    .change_context(errors::ApiErrorResponse::InternalServerError)
-    .attach_printable("Failed to get the connector data")?;
+    // Trigger connector's cancellation
+    } else {
+        // TODO: Remove hardcoded connector
+        let connector_name = api_enums::Connector::Adyen;
+
+        // Form connector data
+        let connector_data: api::ConnectorData = api::ConnectorData::get_connector_by_name(
+            &state.conf.connectors,
+            &connector_name.to_string(),
+            api::GetToken::Connector,
+        )
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Failed to get the connector data")?;
 
-    // Trigger fulfillment
-    let pmd = payouts::PayoutMethodData::default(); // TODO: Fetch from locker
-    payout_data = cancel_payout(
-        state,
-        &merchant_account,
-        &payouts::PayoutRequest::PayoutActionRequest(req.to_owned()),
-        &connector_data,
-        &mut payout_data,
-        &pmd,
-    )
-    .await
-    .attach_printable("Payout fulfillment failed for given Payout request")?;
+        // Trigger fulfillment
+        let payout_method_data = helpers::make_payout_method_data(state, &None, &payout_create)
+            .await?
+            .get_required_value("payout_method_data")?;
```

Comment:
We are passing this in both payout_fulfill and payout_cancel because we have generic functions across all the flows, and this is a mandatory field in PayoutsData which is passed to the connector.

We can remove this dependency by creating separate types for separate actions or making this field Option type